### PR TITLE
Feat/unit test setup

### PR DESCRIPTION
**Pull request by Eysteinn-Orn**
# Units tests and Github Actions for running tests
Adding unit tests and github actions such that when any one of us pushes or pulls to dev we have to pass the unit tests.

# How to test

- run `poetry run pytest` in your console in the root folder of the project.
- automatic linting with GHA

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        id: cache-deps
+        uses: actions/cache@v2
+        with:
+          path: .venv
+          key: pydeps-${{ hashFiles('**/poetry.lock') }}
+
+      - run: poetry install --no-interaction --no-root
+        if: steps.cache-deps.outputs.cache-hit != 'true'
+
+      - run: poetry install --no-interaction
```
Comment (lsig) [inline .github/workflows/test.yml]: Remove this run command, fails because poetry already installed

**Comment by Eysteinn-Orn** [comment]
The GHA seems to be working correctly now but I just need the import to work correctly now in the test file itself

**Comment by Eysteinn-Orn** [comment]
It seems that adding \_\_init\_\_.py to **Every** directory fixes the problem regarding imports and also removing **all** relative imports, additionally we have to rename the types folder since by adding a \_\_init\_\_.py it becomes a package and tries to overwrite an existing python package so I propose to rename that folder to 'typings' but we could also do 'definitions'.

The way I am solving this problem doesn't seem right but it does get the tests running

**Comment by njallskarp** [comment]
Really great addition to the codebase @eysta00, adding a few comments for improvements.

**Review by njallskarp — CHANGES_REQUESTED** [review]
see comments

File: tests/training_test/metrics_test.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+
+from training.metrics import (
+    f1,
+    precision,
+    recall,
+)
+
+
+def test_precision():
```
Comment (njallskarp) [inline tests/training_test/metrics_test.py:11]: two things to improve

1) instead of just testing words.. try to think of more cases. Normal cases such as sentences, what about empty strings? Empty predictions, both empty? Also don't only test edge cases.. test also the 'normal cases'

**Review by njallskarp — CHANGES_REQUESTED** [review]

File: tests/training_test/metrics_test.py (hunk 2 above)
Comment (Eysteinn-Orn) [reply tests/training_test/metrics_test.py:11]: valuable feedback! I will write these improved tests as soon I fix the issue regarding my workflow with GHA, seems like missing GPU libraries on the bot that runs the actions

Comment (njallskarp) [reply tests/training_test/metrics_test.py:11]: OK. Let me know if that remains an issue. I'm curious whether or not it would be possible for us to run the unit tests without needing to use or call any libraries / packages needing a GPU.

Keep me in the loop if you need to discuss this further

Comment (Eysteinn-Orn) [reply tests/training_test/metrics_test.py:11]: so after a little bit of googling and asking AI friends this seems to be the problem;

> Please note that GitHub-hosted runners do not have GPU hardware. Therefore, even if you install the CUDA toolkit, you won't be able to run GPU-accelerated computations on GitHub-hosted runners. If you need to run GPU-accelerated computations as part of your GitHub Actions workflow, you'll need to use self-hosted runners with GPU hardware.

So if we use self-hosted runners we can use GPU hardware there and then test the ML model continuously.

Comment (njallskarp) [reply tests/training_test/metrics_test.py:11]: Is it possible to do the tests without testing the model? If we just test the functions (on a unit level) that should be sufficient for this time being

Comment (Eysteinn-Orn) [reply tests/training_test/metrics_test.py:11]: I think we are already doing that
As it stands we are only running tests on f1, precision and recall. They are all completely independent from the model.

The case might just be that we are importing from the transformers package and that in of itself is causing the problem.

Perhaps we need to decouple the metric functions in metric.py and isolate the evaluate_model function

Comment (njallskarp) [reply tests/training_test/metrics_test.py:11]: Ok I see. I'm pretty sure that transformers should run on a CPU only system. It is possible that they need tensorflow and pytorch as dependencies as well.  Try running the tests on your machine, does that work?

Comment (njallskarp) [reply tests/training_test/metrics_test.py:11]: Additional thought: what happens if you comment out the `evaluate_model` function in metrics.py?

Comment (Eysteinn-Orn) [reply tests/training_test/metrics_test.py:11]: Yes, running the tests locally works without issues :)

Comment (Eysteinn-Orn) [reply tests/training_test/metrics_test.py:11]: Also, I am pretty sure that `poetry install --no-interaction --no-root` should install all our dependencies

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+      - name: Install dependencies
+        if: steps.cached-poetry-dependencies.outputs.cache-hit != 'true'
+        run: poetry install --no-interaction --no-root
+
+      #----------------------------------------------
+      #              run test suite
+      #----------------------------------------------
+      - name: Run tests
+        run: poetry run pytest --rootdir=tests
```
Comment (Eysteinn-Orn) [inline .github/workflows/tests.yml]: Regarding problems with github actions, perhaps writing in this thread would be better

**Comment by Eysteinn-Orn** [comment]
Kinda stuck on this problem of when importing pytorch, no matter what, I get an import error in the github action workflow.

Honestly the way to fix this just momentarily is to decouple the evaluate_model function from metrics.py, but that won't fix the underlying problem of Github Actions having none of the CUDA libraries, been stuck on this for the last two days.

Going to come back to this in the evening and rework the actual tests to just be better.

If we run our own Github Actions server with a GPU then this should be no problem but that is also a whole new can of worms.

**Comment by Eysteinn-Orn** [comment]
Managed to get Pytorch to import correctly using the CPU version of Pytorch, however now there seems to be some version incompatibilities with importing the RobertaForQuestionAnswering transformer
